File: rice-cooker-kotlin/src/test/kotlin/RiceCookerTest.kt
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue


class RiceCookerTest {
	
	private lateinit var riceCooker: RiceCookerModel
	private lateinit var outputStream: ByteArrayOutputStream
	
	@BeforeEach
	fun setUp() {
		riceCooker = RiceCookerModel()
		outputStream = ByteArrayOutputStream()
		System.setOut(PrintStream(outputStream))
	}
	
	@Test
	fun testInitialStatus() {
		assertFalse(riceCooker.isCooking())
		assertEquals("ÉTEINT", riceCooker.getStatus())
	}
	
	@Test
	fun testPowerOnOff() {
		riceCooker.powerOn()
		assertEquals("Le cuiseur de riz est allumé", outputStream.toString().trim())
		
		riceCooker.powerOff()
		assertEquals("Le cuiseur de riz est allumé\n" +
				"Le cuiseur de riz est éteint", outputStream.toString().trim())
	}

	
	@Test
	fun testAddWater() {
		riceCooker.powerOn()
		riceCooker.addWater(3)
		assertEquals("Le cuiseur de riz est allumé\nAjout de 3 litres d'eau. Niveau d'eau actuel : 3 litres", outputStream.toString().trim())
	}
	
	@Test
	fun testCookingFunctionality() {
		riceCooker.powerOn()
		riceCooker.addRice(200)
		riceCooker.cook("10")
		
		assertEquals("Le cuiseur de riz est allumé\n" +
				"Ajout de 200 grammes de riz. Quantité de riz actuelle : 200 grammes\n" +
				"Cuisson pendant 10 minutes", outputStream.toString().trim())
	}
	
	@Test
	fun testBoilFunctionality() {
		riceCooker.powerOn()
		riceCooker.addWater(4)
		riceCooker.boil("5")
		assertEquals("Le cuiseur de riz est allumé\n" +
				"Ajout de 4 litres d'eau. Niveau d'eau actuel : 4 litres\n" +
				"Ébullition pendant 5 minutes", outputStream.toString().trim())
	}
	
	@Test
	fun testKeepingWarmFunctionality() {
		riceCooker.keepWarm()
		assertEquals("Garder le riz au chaud", outputStream.toString().trim())
	}
	
	@Test
	fun testStoppingWarmFunctionality() {
		riceCooker.stopWarm()
		assertEquals("Arrêt de la fonction de maintien au chaud", outputStream.toString().trim())
	}
}

File: rice-cooker-on-scala/src/test/scala/com/ricecooker/RiceCookerTest.scala
package com.ricecooker

import org.scalatest.funsuite.AnyFunSuite

import java.io.{ByteArrayOutputStream, PrintStream}

class RiceCookerTest extends AnyFunSuite {

  test("Test initial status of Rice Cooker") {
    val riceCooker = new RiceCookerModel
    assert(!riceCooker.isCooking)
    assert(!riceCooker.getStatus.contains("allumé"))
  }

  test("Test powering on and off Rice Cooker") {
    val riceCooker = new RiceCookerModel
    riceCooker.powerOn()
    assert(riceCooker.getStatus.contains("allumé"))

    riceCooker.powerOff()
    assert(!riceCooker.getStatus.contains("allumé"))
  }

  test("Test adding water to the Rice Cooker") {
    val riceCooker = new RiceCookerModel

    val outputStream = new ByteArrayOutputStream()
    Console.withOut(new PrintStream(outputStream)) {
      riceCooker.powerOn()
      riceCooker.addWater(100)
    }

    val output = outputStream.toString.trim
    assert(output.contains("Le cuiseur de riz est allumé."))
    assert(output.contains("100 litres d'eau ajoutés."))
  }

  test("Test cooking functionality of the Rice Cooker") {
    val riceCooker = new RiceCookerModel
    val outputStream = new ByteArrayOutputStream()
    Console.withOut(new PrintStream(outputStream)) {
      riceCooker.powerOn()
      riceCooker.addWater(100)
      riceCooker.addRice(200)
      riceCooker.cook("10")
    }

    val output = outputStream.toString.trim
    assert(output.contains("Le cuiseur de riz est allumé."))
    assert(output.contains("100 litres d'eau ajoutés."))
    assert(output.contains("200 grammes de riz ajoutés."))
    assert(output.contains("Cuisson pendant 10 minutes."))
  }

  test("Test boiling water functionality of the Rice Cooker") {
    val riceCooker = new RiceCookerModel
    val outputStream = new ByteArrayOutputStream()
    Console.withOut(new PrintStream(outputStream)) {
      riceCooker.powerOn()
      riceCooker.addWater(100)
      riceCooker.boil("5")
    }

    val output = outputStream.toString.trim
    assert(output.contains("Le cuiseur de riz est allumé."))
    assert(output.contains("100 litres d'eau ajoutés."))
    assert(output.contains("Ébullition d'eau pendant 5 minutes."))
  }

  test("Test keeping warm functionality of the Rice Cooker") {
    val riceCooker = new RiceCookerModel
    val outputStream = new ByteArrayOutputStream()
    Console.withOut(new PrintStream(outputStream)) {
      riceCooker.powerOn()
      riceCooker.keepWarm()
    }

    val output = outputStream.toString.trim

    assert(output.contains("Le cuiseur de riz est allumé."))
    assert(output.contains("Garder le riz au chaud."))
  }

  test("Test when Rice Cooker is not open") {
    val riceCooker = new RiceCookerModel
    val outputStream = new ByteArrayOutputStream()
    Console.withOut(new PrintStream(outputStream)) {
      riceCooker.cook("10")
    }

    val output = outputStream.toString.trim
    assert(output.contains("Impossible de cuisiner. Assurez-vous que le cuiseur de riz est allumé et contient de l'eau et du riz."))
  }
}
